jsonrpc/src/main.ts

```diff
@@ -1090,6 +1124,12 @@ function _createMessageConnection(messageReader: MessageReader, messageWriter: M
 			});
 			if (token) {
 				token.onCancellationRequested(() => {
+					// create cancellation file for the cancel request
+					const cancellationFilename = getCancellationFilename(String(id));
+					if (cancellationFilename) {
+						createCancellationFile(cancellationFilename);
+					}
+
 					connection.sendNotification(CancelNotification.type, { id });
```

Comment:
I want it to work when 2 different type of cancellation strategy is mix used. right now 2 connections are completely independent. otherwise, we need a handshake message when a connection is made to make sure 2 connections are using the same cancellation mode or make master/slave connection to make 1 connection dictate the mode like how client/server works.

Reply:
>  we might cancel twice

for pull mode ("isCancellationRequested"), it is not concern.

for push mode (OnCancellationRequested"), as soon as the event is raised, emitter got removed. so duplicated events can't happen.

and the event source is per request, so how many cancel called on the same source won't matter since multiple cancel call is already guarded as I wrote above.

Reply:
WE can't assume that the other side is implemented in node and therefore uses our implementation of cancellation. So IMO we should only report cancel once. 

I don't see how one connection can use two different cancelation strategies. I always understood that cancellation stragegy is per connection. Either message or file.

Reply:
it is 2 different connections. one used by client and one used by server? it totally make sense from the client/server layer's point of view. but if one see it from jsonrpc or protocol layer's point of view, they can't know?